docs/SILDevelopment.md

```diff
@@ -0,0 +1,159 @@
+for more detailed comments.
+
+## SILType
+
+TBD: Define the different levels of types. Explain type lowering with
+examples.
+
+## SILInstructionResults
+
+TBD: Explain how the various types fit together with pointers to the
+source: SILValue, SILInstruction, SingleValueInstruction,
+MultipleValueInstructionResult. And why it was done this way.
+
+## `SILFunction` and `apply` arguments.
+
+Throughout the compiler, integer indices are used to identify argument
+positions in several different contexts:
+
+- `SILFunctionType` parameters.
+
+- The SIL function definition's callee-side `SILFunctionArguments`,
+  including indirect results.
+
+- Caller-side "applied arguments" for `apply` and `try_apply` (vs. operands).
+
+- Captured "applied arguments" for `partial_apply` (vs. operands).
+
+- SIL arguments (any of the three above) at the Lowered SIL stage.
+
+An argument may have a different position in each of these
+contexts. It should be obvious that code should never assume and
+hard-code argument positions or offsets. Furthermore, common idioms
+for accessing arguments should be adopted throughout the
+compiler. Plenty of bugs have resulted from assumptions about the form
+of SIL made in one area of the compiler that have been copied into
+other parts of the compiler. Also, without consistent use of common
+idioms, it becomes overly burdensome to evolve these APIs over time.
+
+> Note: Hardcoding offsets based on assumptions of operand layout are still
+> surprisingly prevalent in the SIL implementation. This was carried over from
+> LLVM practice but is actually a serious problem in SIL because: (a) the
+> structure of SIL instructions, particularly applies, is much more nuanced than
+> LLVM IR, (b) assumptions that may be valid at the initial point of use are
+> often copied into parts of the code where they are no longer valid; and (c)
+> unlike LLVM IR, SIL is not stable and will continue evolving.
+
+```
+func example<T>(t: T) -> T {
+  let foo = { return t }
+  return foo()
+}
+```
+
+Ignoring conventions, the closure `foo` would map to the following indices...
+
+One direct formal parameter and one direct formal result:
+
+```
+SILFunctionType(foo):  $(0:T) -> @out T
+```
+
+Canonical SIL with opaque values matches `SILFunctionType` with one
+direct `SILFunctionArgument` and one direct SILResult:
+
+```
+SILFunctionArguments:  (0: $T) -> $T
+```
+
+Lowered SIL has two indirect SILFunctionArguments, the first of which
+is now a "result argument":
+
+```
+SILFunctionArguments:  (0: $*T, 1: $*T)
+```
+
+Creation of the closure has one applied argument:
+
+```
+%closure = partial_apply @foo(0: %t)
+```
+
+Application of the closure with opaque values has zero applied arguments:
+
+```
+%result = apply %closure()
+```
+
+Lowered application of the closure has one applied argument:
+
+```
+apply %closure(%result: $*T)
+```
+
+The mapping between SILFunctionType and SILFunctionArguments depends
```

Comment:
Maybe move this before the previous section? Then it would be clearer what you are trying to show.